File: hourrank-5/xor-se/xor_se.c
#include<stdio.h>
#include<stdlib.h>

// https://www.hackerrank.com/contests/hourrank-5/challenges/xor-se

long int xor_sum(long int L, long int R) {
    int ones = 1;
    long s = 0;
    long start = L;
    long stop = R;
    long i = 0;
    int m = 0;

    printf("start = %ld, stop = %ld\n",L,R);

    if (start%2 != 0) {
        s = s ^ start;
        start += 1;
    }

    if (stop%2 == 0) {
        s = s ^ stop;
        stop = R-1;
    }

    for (i = start; i <= stop; i+=2) {
        printf("iterating %ld\n",i);
        m = i&3;
        if (m == 0) {
            s = s ^ i;
        } else if (m == 2) {
            s = s ^ (i+1);
        }
    }

    return s;
}

int main() {

    int a = 0;
    int Q = 0;
    long *ranges = NULL;
    long s = 0;

    scanf("%d",&Q);

    ranges = (long*) calloc(2*Q,sizeof(Q));

    for (a = 0; a < 2*Q; a+=2) {
        scanf("%ld %ld", &ranges[a], &ranges[a+1]);
    }

    for (a = 0; a < 2*Q; a+=2) {
        s = xor_sum(ranges[a],ranges[a+1]);
        printf("%ld",s);
    }

    free(ranges);

    return 0;
}

File: hourrank-5/xor-se/xor_se.py
# https://www.hackerrank.com/contests/hourrank-5/challenges/xor-se

def xor_sum(L,R):
    """return the xor sum of the elements in the sequence from L to R
    """
    s = 0
    print L,R

    final_xors = []

    # 1^1 == 0
    # count the number of 1's and xor by 1 at the end if needed
    # the number of 1's is R/4
    # default to 1 for now, fix later
    final_xors.append(1)

    start = L
    stop = R+1

    # start automatic xor'ing on an even value
    if start%2 != 0:
        final_xors.append(L)
        start += 1

    # end automatic xor'ing on an even value
    # stop is even, means R is odd
    if stop%2 == 0:
        final_xors.append(R)
        stop = R

    for i in xrange(start, stop, 2):
        m = i&3
        if m == 0:
            s = s ^ i
        elif m == 2:
            s = s ^ (i+1)

    for i in final_xors:
        s = s ^ i

    return s


if __name__ == '__main__':
    Q = int(raw_input().strip())

    inputs = []

    for a0 in xrange(Q):
        L,R = raw_input().strip().split(' ')
        inputs.append([long(L),long(R)])


#    for a in xrange(Q):
#        s = xor_sum(*inputs[a])
#        print s

    for m1 in xrange(11):
        for m2 in xrange(11):
            xor_sum()
